etc/benchmark-uploads/main.go

```diff
@@ -0,0 +1,381 @@
+	cancelCh := make(chan os.Signal, 1)
+	signal.Notify(cancelCh, os.Interrupt)
+	go func() {
+		<-cancelCh
+		cancel()
+		signal.Stop(cancelCh)
+		close(cancelCh)
+	}()
+
+	// Setup pach context for creating destination repo, etc.
+	c, err := client.NewOnUserMachine("")
+	if err != nil {
+		log.Fatal(err)
+	}
+	c = c.WithCtx(ctx)
+	if err := c.CreateProjectRepo("default", "benchmark-upload"); err != nil {
+		log.Printf("create repo: %v", err)
+	}
+
+	hc := &http.Client{}
+	if *h2c {
+		hc = &http.Client{
+			Transport: &http2.Transport{
+				AllowHTTP: true,
+				DialTLSContext: func(ctx context.Context, network, addr string, cfg *tls.Config) (net.Conn, error) {
+					return (&net.Dialer{}).DialContext(ctx, network, addr)
+				},
+			},
+		}
+	}
+
+	// Run benchmark.
+	commit := client.NewProjectCommit("default", "benchmark-upload", "master", "")
+	var benchErr error
+	switch *kind {
+	case "grpc":
+		benchErr = bench(func(name string, r io.Reader, _ uint64) error {
+			return c.PutFile(commit, name, r)
+		})
+	case "http":
+		benchErr = bench(func(name string, r io.Reader, _ uint64) error {
+			req, err := http.NewRequestWithContext(ctx, "PUT", *scheme+"://"+*addr+"/upload/"+name, r)
+			if err != nil {
+				return err
+			}
+			res, err := hc.Do(req)
+			if err != nil {
+				return err
+			}
+			defer res.Body.Close()
+
+			d, err := httputil.DumpResponse(res, true)
+			if err != nil {
+				return err
+			}
+			log.Printf("%s", d)
+
+			if got, want := res.StatusCode, http.StatusAccepted; got != want {
+				return fmt.Errorf("unexpected status: got %d want %d: %v", got, want, res.Status)
+			}
+			return nil
+		})
+	case "console":
+		benchErr = bench(func(name string, r io.Reader, length uint64) error {
+			// Start
+			start, err := json.Marshal(map[string]any{
+				"branch": "master",
+				"path":   "/",
+				"repo":   "benchmark-upload",
+			})
+			if err != nil {
+				return errors.Wrap(err, "marshal start")
+			}
+			req, err := http.NewRequestWithContext(ctx, "POST", *scheme+"://"+*addr+"/upload/start", bytes.NewReader(start))
+			if err != nil {
+				return err
+			}
+			req.Header.Add("content-type", "application/json")
+			req.AddCookie(&http.Cookie{
+				Name:  "dashAuthToken",
+				Value: c.AuthToken(),
+			})
+			res, err := hc.Do(req)
+			if err != nil {
+				return errors.Wrap(err, "start")
+			}
+			body, err := io.ReadAll(res.Body)
+			res.Body.Close()
+			if err != nil {
+				return errors.Wrap(err, "read start body")
+			}
+			if got, want := res.StatusCode, http.StatusOK; got != want {
+				return fmt.Errorf("start: unexpected status code: got %d want %d %v", got, want, res.Status)
+			}
+			var idInfo struct {
+				UploadId string `json:"uploadId"`
+			}
+			if err := json.Unmarshal(body, &idInfo); err != nil {
+				return errors.Wrap(err, "unmarshal start reply")
+			}
+
+			// Upload
+			ckSize := 50 * units.MB
+			nCk := int(math.Ceil(float64(length) / float64(ckSize)))
```

Comment:
not really important, but if you want `ceil` without the float conversion, you can do `(length + ckSize - 1)/ckSize`

If you're interested: http://blog.pkh.me/p/36-figuring-out-round%2C-floor-and-ceil-with-integer-division.html and in particular `⌈n/m⌉ = ⌊(n-1)/m⌋+1 = ⌊(n+m-1)/m⌋`

Reply:
That is a really good tip.  I always feel dirty doing the float conversion and never stopped to think if there was a better way of doing that.  I'll have to add this to my muscle memory!